lib/maths/CBoostedTreeLoss.cc

```diff
@@ -366,6 +366,144 @@ CArgMinMultinomialLogisticLossImpl::objectiveGradient() const {
+}
+
+bool CArgMinMsleImpl::nextPass() {
+    ++m_CurrentPass;
+    return this->bucketWidth().first > 0.0 && m_CurrentPass < 2;
+}
+
+void CArgMinMsleImpl::add(const TMemoryMappedFloatVector& prediction, double actual, double weight) {
+    double expPrediction{std::exp(prediction[0])};
+    double logActual{CTools::fastLog(1.0 + actual)};
+    switch (m_CurrentPass) {
+    case 0: {
+        m_ExpPredictionMinMax.add(expPrediction);
+        m_LogActualMinMax.add(logActual);
+        m_MeanLogActual.add(logActual, weight);
+        break;
+    }
+    case 1: {
+        double logError{logActual - CTools::fastLog(1.0 + expPrediction)};
+
+        TVector example;
+        example(0) = expPrediction;
+        example(1) = logActual;
+        example(2) = logError;
+        auto bucketIndex{this->bucket(expPrediction, logActual)};
+        m_Buckets[bucketIndex.first][bucketIndex.second].add(example, weight);
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void CArgMinMsleImpl::merge(const CArgMinLossImpl& other) {
+    const auto* mlse = dynamic_cast<const CArgMinMsleImpl*>(&other);
+    if (mlse != nullptr) {
+        switch (m_CurrentPass) {
+        case 0:
+            m_ExpPredictionMinMax += mlse->m_ExpPredictionMinMax;
+            m_LogActualMinMax += mlse->m_LogActualMinMax;
+            m_MeanLogActual += mlse->m_MeanLogActual;
+            break;
+        case 1:
+            for (std::size_t i = 0; i < m_Buckets.size(); ++i) {
+                for (std::size_t j = 0; j < m_Buckets[i].size(); ++j) {
+                    m_Buckets[i][j] += mlse->m_Buckets[i][j];
+                }
+            }
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+CArgMinMsleImpl::TDoubleVector CArgMinMsleImpl::value() const {
+    std::function<double(double)> objective;
+    double minLogWeight;
+    double maxLogWeight;
+
+    objective = this->objective();
+    if (this->bucketWidth().first == 0.0) {
+        minLogWeight = -4.0;
+        maxLogWeight = CBasicStatistics::mean(m_MeanLogActual);
+    } else {
+        // If the weight smaller than the minimum log error every prediction is low.
+        // Conversely, if it's larger than the maximum log error every prediction is
+        // high. In both cases, we can reduce the error by making the weight larger,
+        // respectively smaller. Therefore, the optimal weight must lie between these
+        // values.
+        minLogWeight = std::numeric_limits<double>::max();
+        maxLogWeight = -minLogWeight;
+        for (const auto& bucketsPrediction : m_Buckets) {
+            for (const auto& bucketActual : bucketsPrediction) {
+                minLogWeight = std::min(minLogWeight,
+                                        CBasicStatistics::mean(bucketActual)(2));
+                maxLogWeight = std::max(maxLogWeight,
+                                        CBasicStatistics::mean(bucketActual)(2));
```

Comment:
I wonder if it might be worthwhile naming constants, i.e. `static constexpr std::size_t PREDICTION = 0`, ..., `static constexpr std::size_t ERROR = 2`. When I first read this I thought it was an error, but of course you're actually reading the error out of the bucket, so doing the right thing if this was it would make this clearer I think
```suggestion
                                        CBasicStatistics::mean(bucketActual)(ERROR));
```

Reply:
Fixed